### PR TITLE
Github Deployment Watchdog Worker

### LINKED ISSUE #96: DM user if no deployment starts in X seconds
Sometimes, you might /deploy and nothing happens. There's a few things that could cause this:

1. The repository doesn't have a webhook configured to handle the `deployment` event.
2. The service that handles the `deployment` event (and subsequently posts back `deployment_status` events) is broken.
3. GitHub is down, or otherwise backed up on delivering webhooks

In either case, a simple DM back to the user to let them know that nothing has happened yet, and some helpful steps to debug ("check status.github.com", "does your repo have a webhook to handle github deployments"?)

### PR DESCRIPTION
**Pull request by russellballestrini**
The Github Deployment Watchdog queues up when a create_deployment is requested from github. It wakes up after 30 seconds to see if there are any Github Deployment Statuses associated with the Gihub Deployment. If there are none, it will send a notification to the slack user letting them know something is wrong, otherwise it exits happily and does nothing.

Fixes #96 

	modified:   Procfile
	new file:   app/messages/github_no_deployment_status_message.rb
	modified:   app/models/commit_status_context.rb
	modified:   app/models/deployment.rb
	new file:   app/models/deployment_status.rb
	modified:   app/views/messages/auto_deployment_stuck_pending.text.erb
	new file:   app/views/messages/github_no_deployment_status.text.erb
	renamed:    app/workers/deployment_watchdog_worker.rb -> app/workers/auto_deployment_watchdog_worker.rb
	new file:   app/workers/github_deployment_watchdog_worker.rb
	modified:   lib/github/client/octokit.rb
	modified:   lib/slashdeploy/service.rb
	modified:   spec/features/auto_deploy_spec.rb
	modified:   spec/github/client/octokit_spec.rb

### CODE REVIEW

File: app/views/messages/github_no_deployment_status.text.erb [hunk 1]
```diff
@@ -0,0 +1 @@
+:sadparrot: <@<%= @account.id %>>, The Github Deployment <<%= @github_deployment.id %>> of <%= @github_deployment.repository %>@<<%= @github_deployment.sha %>|<%= @github_deployment.ref %>> to *<%= @github_deployment.environment %>* did _not_ start. Please review this repo's `.slashdeploy.yml` file and/or have @OpsEng review this repo's Empire Github App Integration Webhooks (for more details, visit: https://remind.quip.com/UbN0AaMDsn8N#dcPACAHqhCJ)
```
Comment (ejholmes) [inline app/views/messages/github_no_deployment_status.text.erb]: SlashDeploy is open source, so messages with content pointing to internal Remind docs is a no go.

Let's update `app/views/documentation/index.html.erb` with some debugging information about the common cases that can result in this scenario, and then change this to link to https://slashdeploy.io/docs:

1. Repository doesn't have a webhook handling the `deployment` event. Or..
2. The service handling the `deployment` event isn't updating GitHub with Deployment Statuses.
   * Service may be hung/down/etc.
   * Service may have ignored the deployment request, if it's filtering for a specific environment (e.g. it'll deploy to `production`, but that provided environment in the `deployment` event was `prod`).

File: app/models/deployment.rb [hunk 2]
```diff
@@ -1,17 +1,32 @@
...
+  end
+
+  # memoize the parts of this deployment's url.
+  def url_parts
+    # ["https:", "", "api.github.com", "repos", "octocat", "example", "deployments", "1"]
+    @parts ||= url.split("/").to_a
+  end
+
+  # return the Github Org for deployment.
+  def organization
```
Comment (ejholmes) [inline app/models/deployment.rb]: This seems a little error prone. Let's follow the same pattern we do in the `Repository` model:

https://github.com/remind101/slashdeploy/blob/2dd90f45b07affafe2e91f34b6f715201e6a2f2e/app/models/repository.rb#L49-L52

And match on `self.repository`.

**Comment by russellballestrini** [comment]
Hey @ejholmes this is ready for another review.

File: app/views/documentation/index.html.erb [hunk 3]
```diff
@@ -74,5 +86,23 @@ production:
...
       <dd>When you first run the /deploy command, we ask you to authenticate with your GitHub account, which is then linked to your Slack user account.</dd>
     </dl>
+
+    <hr />
+
+    <h5 id="troubleshooting">Troubleshooting</h5>
+    <p>Here we have some debugging information about the common issues and how to resolve them.
+    <dl>
+      <dt id="error-1">Error 1: Deployment did not start</dt>
+      <dd>If a deployment doesn't start with in 30 seconds, you will get a Slack notification letting you know something went wrong. This might happen for a number of reasons, so check the following.
```
Comment (ejholmes) [inline app/views/documentation/index.html.erb]: Small typo: `/with in/within`

Comment (ejholmes) [reply app/views/documentation/index.html.erb]: Also, maybe replace `30` with `<%= time_ago_in_words(GithubDeploymentWatchdogWorker::DEFAULT_DELAY.from_now, include_seconds: true) %>`

File: app/views/documentation/index.html.erb [hunk 4]
```diff
@@ -74,5 +86,23 @@ production:
...
+      <ul>
+        <li>make sure the repo's <code>slashdeploy.yml</code> file is configured properly</li>
+        <li>make sure the repo has a Github Deployment Integration (and Webhooks) to handle deployment events</li>
+        <li>make sure the service handling the deployment events is working and sending Deployment Status updates to Github</li>
+        <li>make sure the service handling the deployment events is not filtering or ignoring deployment requests. For example, it will deploy to <code>production</code>, however the  provided environment in the deployment event was <code>prod</code>.
+      </dd>
+      <dt id="error-2">Error 2: Required commit status contexts stuck in pending</dt>
+      <dd>Inside the repo's <code>slashdeploy.yml</code> a list of <code>required_contexts</code> was defined which need to move into the <code>success</code> state in order to continue. For some reason, after an hour of waiting, at least one of these <code>required_contexts</code> are stuck in the <code>pending</code> state instead of transitioning to <code>success<code>, <code>failure</code>, or <code>error</code>. If you fix this issue your deployment will likely continue.</dd>
```
Comment (ejholmes) [inline app/views/documentation/index.html.erb]: Maybe change `after an hour` to `after <%= time_ago_in_words(DeploymentWatchdogWorker::DEFAULT_DELAY.from_now) %>`

File: app/workers/github_deployment_watchdog_worker.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  # notify the user if there was an issue otherwise do nothing.
+  # user_id: the database id of the slashdeploy User object.
+  # github_repo: the repo of the Github Deployment.
+  # github_deployment_id: the id of the external Github Deployment.
+  def perform(user_id, github_repo, github_deployment_id)
+    # get the slashdeploy User object.
+    user = User.find(user_id)
+
+    # get the Github Deployment by id.
+    github_deployment = SlashDeploy.service.github.get_deployment(user, github_repo, github_deployment_id)
```
Comment (ejholmes) [inline app/workers/github_deployment_watchdog_worker.rb]: `SlashDeploy::Service#github` is assumed private (we should actually explicitly flag it as private). Let's wrap these two lines up in a method on `SlashDeploy::Service`:

```ruby
# Returns truthy if the given deployment has started (has deployment statuses)
def deployment_started?(user, repo, deployment_id)
  deployment = github.get_deployment(user, repo, deployment_id)
  deployment_status = github.last_deployment_status(user, github_deployment.url)
  deployment if deployment_status
end
```

File: spec/features/commands_spec.rb [hunk 6]
```diff
@@ -534,6 +534,41 @@
...
+      [users(:david), DeploymentRequest.new(repository: 'acme-inc/api', ref: 'master', environment: 'production')]
+    ]
+
+    # our deployment watchdog worker should start with an empty queue.
+    expect(GithubDeploymentWatchdogWorker.jobs.size).to eq 1
+
+    # setup expectations of the worker notifying the user there was an issue.
+    expect(SlashDeploy.service).to receive(:direct_message).with(
+      slack_accounts(:david),
+      GithubNoDeploymentStatusMessage,
```
Comment (ejholmes) [inline spec/features/commands_spec.rb]: We should be asserting the message content:

```ruby
expect(slack).to receive(:direct_message).with(
  slack_accounts(:david),
  Slack::Message.new(...)
)
```

It's somewhat verbose, but has saved me a bunch of times during refactors (we can cut down on the verbosity in the future by using something like gold master approvals).

File: spec/features/commands_spec.rb [hunk 7]
```diff
@@ -534,6 +534,41 @@
...
   end
 
+  scenario 'github deployment does not start after 30 simulated secs and triggers watchdog' do
+    # make sure our queue is clear before starting test.
+    GithubDeploymentWatchdogWorker.clear
+
+    # our deployment watchdog worker should start with an empty queue.
+    expect(GithubDeploymentWatchdogWorker.jobs.size).to eq 0
+
+    expect(User).to receive(:find).and_return(users(:david))
```
Comment (ejholmes) [inline spec/features/commands_spec.rb]: No need to stub this. These tests shouldn't mock out the database.

File: spec/features/commands_spec.rb [hunk 8]
```diff
@@ -534,6 +534,41 @@
...
+    expect(users(:david)).to receive(:slack_account_for_github_organization).and_return(slack_accounts(:david))
```
Comment (ejholmes) [inline spec/features/commands_spec.rb]: Also no need to stub this.

File: spec/features/commands_spec.rb (hunk 6 above)
Comment (russellballestrini) [reply spec/features/commands_spec.rb]: @ejholmes        `undefined local variable or method `slack' for #<RSpec::ExampleGroups::SlashCommands_2:0x007fdf72e66eb0>`

File: spec/features/commands_spec.rb [hunk 9]
```diff
@@ -534,6 +534,38 @@
...
+    expect(command_response).to be_in_channel
+    expect(deployment_requests).to eq [
+      [users(:david), DeploymentRequest.new(repository: 'acme-inc/api', ref: 'master', environment: 'production')]
+    ]
+
+    # our deployment watchdog worker should start with an empty queue.
+    expect(GithubDeploymentWatchdogWorker.jobs.size).to eq 1
+
+    # setup expectations of the worker notifying the user there was an issue.
+    expect(SlashDeploy.service.slack).to receive(:direct_message).with(
```
Comment (ejholmes) [inline spec/features/commands_spec.rb]: You can just use `expect(slack)` here.

Comment (russellballestrini) [reply spec/features/commands_spec.rb]: I got an exception

**Pull request merged**